src/stransport_stream.c

```diff
@@ -150,9 +156,9 @@ static OSStatus read_cb(SSLConnectionRef conn, void *data, size_t *len)
 	requested = left = *len;
 	do {
 		ret = git_stream_read(io, data + (requested - left), left);
-		if (ret < 0) {
+		if (ret <= 0) { /* There was an error or nothing could be read because stream is closed */
 			*len = 0;
-			return -1;
+			return errSSLClosedAbort;
```

Comment:
BTW if you look at the use of this very API by Apple at (http://www.opensource.apple.com/source/Security/Security-55471/libsecurity_ssl/regressions/ssl-45-tls12.c), you will notice `SocketRead()` is returning an error when reading 0 bytes. And both `SocketRead()~ and `SocketWrite()` use loops to read / write exactly the passed-in number of bytes.